Error while connecting with TLS certificates

I'm trying to connect to the server with the self-signed TLS client certificates (created with [generate-CA.sh](https://raw.githubusercontent.com/owntracks/tools/master/TLS/generate-CA.sh)), but I'm having some issues.

```
2022-06-05 22:51:15.979 D MessageProcessor: message:TLS setup failed: error constructing MAC: java.security.InvalidKeyException: No installed provider supports this key: com.android.org.bouncycastle.jcajce.PKCS12Key, 
2022-06-05 22:51:15.980 D BackgroundService: endpoint state changed Error. Message: TLS setup failed: error constructing MAC: java.security.InvalidKeyException: No installed provider supports this key: com.android.org.bouncycastle.jcajce.PKCS12Key
java.io.IOException: error constructing MAC: java.security.InvalidKeyException: No installed provider supports this key: com.android.org.bouncycastle.jcajce.PKCS12Key
	at com.android.org.bouncycastle.jcajce.provider.keystore.pkcs12.PKCS12KeyStoreSpi.engineLoad(PKCS12KeyStoreSpi.java:852)
	at java.security.KeyStore.load(KeyStore.java:1484)
	at org.owntracks.android.support.SocketFactory.<init>(SourceFile:304)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.getMqttConnectOptions(SourceFile:186)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.connectToBroker(SourceFile:177)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.sendMessage(SourceFile:30)
	at org.owntracks.android.services.MessageProcessor.lambda$sendAvailableMessages$0(Unknown Source:2)
	at org.owntracks.android.services.MessageProcessor.$r8$lambda$bVYaIhYhXtJCzBaSxWOGtcUu6f8(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessor$$ExternalSyntheticLambda1.call(Unknown Source:4)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.os.HandlerThread.run(HandlerThread.java:67)
```

Here's what I've tried so far:
- Turn off TLS and verify that the recorder config is correctly set up
- Used `mosquitto_pub` and `mosquitto_sub` to verify that certs generated with `generate-CA.sh` are working fine
- Reinstall the app
- Tried with new users/certs
- Verified that I'm using the same CA for the server/client cert.

...but no luck so far. Not sure what's wrong, but my only guess is that either `openssl` is generating incorrect `.p12` file or the OS/app somehow can't find the API required to open PKCS12 file.

Here are some additional information that can help the debugging process: 
1. I've used the following command to create `.p12` file from the existing crt/key file. `openssl pkcs12 -export -in name.crt -inkey name.key -name "name-cert"  -out name.p12`
2. When I tried to add `name.p12` file in the system settings, it did not accept the password I put in. It's a very simple password and I've verified it multiple times. Perhaps this is somehow related?
* **App build number**: 2.4.8, 20408019
* **Android version**: 12
* **Device**: Galaxy S21 Ultra
* **Installation source**: Google play

Thanks for all the help!

Hi mate,

it actually looks like there is something wrong with the certificate,

This is what I get in owntracks and from my understanding it is the reject from the mqtt server due to the client certificate is not trusted,
```
2022-11-03 07:04:05.837 I TimberInMemoryLogTree: Logs cleared
2022-11-03 07:04:23.089 I ServiceStarter$Impl: starting service
2022-11-03 07:04:35.630 E MessageProcessorEndpointMqtt: Failed to reconnect to MQTT broker
org.owntracks.android.services.MqttConnectionException:  (0) - javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb4000073695c6208: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736940e188:0x00000001)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.connectToBroker(SourceFile:243)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.reconnect(SourceFile:46)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.lambda$reconnect$1(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.$r8$lambda$3UR4gQL2b_UbF_ekJLIMhafolpU(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt$$ExternalSyntheticLambda0.run(Unknown Source:18)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:233)
	at android.os.Looper.loop(Looper.java:344)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by:  (0) - javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb4000073695c6208: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736940e188:0x00000001)
	at androidx.work.impl.utils.WorkForegroundUpdater$1.run(SourceFile:433)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb4000073695c6208: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736940e188:0x00000001)
	at org.conscrypt.SSLUtils.toSSLHandshakeException(Unknown Source:13)
	at org.conscrypt.ConscryptEngine.convertException(Unknown Source:14)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:369)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:28)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:11)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(SourceFile:70)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.access$100(Unknown Source:0)
	at org.conscrypt.ConscryptEngineSocket.doHandshake(Unknown Source:90)
	at org.conscrypt.ConscryptEngineSocket.startHandshake(Unknown Source:36)
	at org.eclipse.paho.client.mqttv3.internal.SSLNetworkModule.start(Unknown Source:83)
	at androidx.work.impl.utils.WorkForegroundUpdater$1.run(SourceFile:174)
	... 1 more
Caused by: javax.net.ssl.SSLProtocolException: Read error: ssl=0xb4000073695c6208: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736940e188:0x00000001)
	at org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
	at org.conscrypt.NativeSsl.readDirectByteBuffer(Unknown Source:16)
	at org.conscrypt.ConscryptEngine.readPlaintextDataDirect(Unknown Source:6)
	at org.conscrypt.ConscryptEngine.readPlaintextData(Unknown Source:21)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:253)
	... 9 more

2022-11-03 07:04:37.854 I MQTTReconnectWorker: MQTTReconnectWorker started on threadID: Thread[androidx.work-2,5,main]
2022-11-03 07:04:38.048 E MessageProcessorEndpointMqtt: Failed to reconnect to MQTT broker
org.owntracks.android.services.MqttConnectionException:  (0) - javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb40000736a9bf488: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736953a5c8:0x00000001)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.connectToBroker(SourceFile:243)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.reconnect(SourceFile:46)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.lambda$reconnect$1(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.$r8$lambda$3UR4gQL2b_UbF_ekJLIMhafolpU(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt$$ExternalSyntheticLambda0.run(Unknown Source:18)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:233)
	at android.os.Looper.loop(Looper.java:344)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by:  (0) - javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb40000736a9bf488: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736953a5c8:0x00000001)
	at androidx.work.impl.utils.WorkForegroundUpdater$1.run(SourceFile:433)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb40000736a9bf488: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736953a5c8:0x00000001)
	at org.conscrypt.SSLUtils.toSSLHandshakeException(Unknown Source:13)
	at org.conscrypt.ConscryptEngine.convertException(Unknown Source:14)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:369)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:28)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:11)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(SourceFile:70)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.access$100(Unknown Source:0)
	at org.conscrypt.ConscryptEngineSocket.doHandshake(Unknown Source:90)
	at org.conscrypt.ConscryptEngineSocket.startHandshake(Unknown Source:36)
	at org.eclipse.paho.client.mqttv3.internal.SSLNetworkModule.start(Unknown Source:83)
	at androidx.work.impl.utils.WorkForegroundUpdater$1.run(SourceFile:174)
	... 1 more
Caused by: javax.net.ssl.SSLProtocolException: Read error: ssl=0xb40000736a9bf488: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb40000736953a5c8:0x00000001)
	at org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
	at org.conscrypt.NativeSsl.readDirectByteBuffer(Unknown Source:16)
	at org.conscrypt.ConscryptEngine.readPlaintextDataDirect(Unknown Source:6)
	at org.conscrypt.ConscryptEngine.readPlaintextData(Unknown Source:21)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:253)
	... 9 more

2022-11-03 07:05:14.794 W MessageProcessorEndpointMqtt: failed connection attempts: 21
2022-11-03 07:05:14.794 E MessageProcessor: Message delivery failed. queueLength: 2, messageId: 1667455267380-c36bb2
2022-11-03 07:05:14.794 W MessageProcessor: Error sending message. Re-queueing
2022-11-03 07:05:14.797 I MessageProcessor: Waiting for 120 s before retrying
2022-11-03 07:05:48.071 I MQTTReconnectWorker: MQTTReconnectWorker started on threadID: Thread[androidx.work-3,5,main]
2022-11-03 07:05:48.404 E MessageProcessorEndpointMqtt: Failed to reconnect to MQTT broker
org.owntracks.android.services.MqttConnectionException:  (0) - javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb4000073692e8408: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb4000073692ac388:0x00000001)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.connectToBroker(SourceFile:243)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.reconnect(SourceFile:46)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.lambda$reconnect$1(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt.$r8$lambda$3UR4gQL2b_UbF_ekJLIMhafolpU(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessorEndpointMqtt$$ExternalSyntheticLambda0.run(Unknown Source:18)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:233)
	at android.os.Looper.loop(Looper.java:344)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by:  (0) - javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb4000073692e8408: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb4000073692ac388:0x00000001)
	at androidx.work.impl.utils.WorkForegroundUpdater$1.run(SourceFile:433)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: javax.net.ssl.SSLHandshakeException: Read error: ssl=0xb4000073692e8408: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb4000073692ac388:0x00000001)
	at org.conscrypt.SSLUtils.toSSLHandshakeException(Unknown Source:13)
	at org.conscrypt.ConscryptEngine.convertException(Unknown Source:14)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:369)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:28)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:11)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(SourceFile:70)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.access$100(Unknown Source:0)
	at org.conscrypt.ConscryptEngineSocket.doHandshake(Unknown Source:90)
	at org.conscrypt.ConscryptEngineSocket.startHandshake(Unknown Source:36)
	at org.eclipse.paho.client.mqttv3.internal.SSLNetworkModule.start(Unknown Source:83)
	at androidx.work.impl.utils.WorkForegroundUpdater$1.run(SourceFile:174)
	... 1 more
Caused by: javax.net.ssl.SSLProtocolException: Read error: ssl=0xb4000073692e8408: Failure in SSL library, usually a protocol error
error:10000418:SSL routines:OPENSSL_internal:TLSV1_ALERT_UNKNOWN_CA (/usr/local/google/home/prb/src/boringssl/ssl/tls_record.cc:594 0xb4000073692ac388:0x00000001)
	at org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
	at org.conscrypt.NativeSsl.readDirectByteBuffer(Unknown Source:16)
	at org.conscrypt.ConscryptEngine.readPlaintextDataDirect(Unknown Source:6)
	at org.conscrypt.ConscryptEngine.readPlaintextData(Unknown Source:21)
	at org.conscrypt.ConscryptEngine.unwrap(Unknown Source:253)
	... 9 more

```

I get the following error on the mqtt server
```
OpenSSL Error: error:140890B2:SSL routines:SSL3_GET_CLIENT_CERTIFICATE:no certificate returned
 Socket error on client <unknown>, disconnecting.
```

For some reason I can not use openssl to check / extract the certificate even if I paste the password,
```
openssl pkcs12 -info -in test-client.p12 -nodes
Enter Import Password:
MAC: sha1, Iteration 1
MAC length: 20, salt length: 8
PKCS7 Encrypted data: pbeWithSHA1And40BitRC2-CBC, Iteration 2048
Error outputting keys and certificates
801B94A8707F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:../crypto/evp/evp_fetch.c:349:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
```

When I use the ubuntu gui to check the certificate (pasting the same password), I can see that there is a CA included in your certificate. I will try to see if that would help in my case.

--- a/project/app/build.gradle.kts
+++ b/project/app/build.gradle.kts
@@ -214,6 +214,8 @@ dependencies {
     implementation(libs.apache.httpcore)
     implementation(libs.commons.codec)
 
+    implementation(libs.bouncycastle)
+
     // Widget libraries
     implementation(libs.widgets.materialedittext)
     implementation(libs.widgets.materialdrawer) { artifact { type = "aar" } }

--- a/project/app/build.gradle.kts
+++ b/project/app/build.gradle.kts
@@ -214,6 +214,8 @@ dependencies {
     implementation(libs.apache.httpcore)
     implementation(libs.commons.codec)
 
+    // The BC version shipped under com.android is half-broken. Weird certificate issues etc.
+    // To solve, we bring in our own version of BC
     implementation(libs.bouncycastle)
 
     // Widget libraries

--- a/project/app/src/main/java/org/owntracks/android/App.kt
+++ b/project/app/src/main/java/org/owntracks/android/App.kt
@@ -70,6 +70,8 @@ class App : Application(), Configuration.Provider {
                 Conscrypt.newProviderBuilder().provideTrustManager(false).build(), 1
             )
         }
+
+        // Bring in a real version of BC and don't use the device version.
         Security.removeProvider("BC")
         Security.addProvider(BouncyCastleProvider())
 

--- a/project/app/src/main/java/org/owntracks/android/App.kt
+++ b/project/app/src/main/java/org/owntracks/android/App.kt
@@ -15,6 +15,7 @@ import androidx.work.Configuration
 import androidx.work.WorkerFactory
 import dagger.hilt.EntryPoints
 import dagger.hilt.android.HiltAndroidApp
+import org.bouncycastle.jce.provider.BouncyCastleProvider
 import org.conscrypt.Conscrypt
 import org.owntracks.android.di.CustomBindingComponentBuilder
 import org.owntracks.android.di.CustomBindingEntryPoint
@@ -69,6 +70,8 @@ class App : Application(), Configuration.Provider {
                 Conscrypt.newProviderBuilder().provideTrustManager(false).build(), 1
             )
         }
+        Security.removeProvider("BC")
+        Security.addProvider(BouncyCastleProvider())
 
         super.onCreate()
 

--- a/project/app/src/main/java/org/owntracks/android/support/SocketFactory.java
+++ b/project/app/src/main/java/org/owntracks/android/support/SocketFactory.java
@@ -8,6 +8,7 @@ import java.security.KeyManagementException;
 import java.security.KeyStore;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
+import java.security.Security;
 import java.security.UnrecoverableKeyException;
 import java.security.cert.CertificateFactory;
 import java.security.cert.X509Certificate;
@@ -113,7 +114,7 @@ public class SocketFactory extends javax.net.ssl.SSLSocketFactory{
         if (options.hasClientP12Crt()) {
             Timber.v("options.hasClientP12Crt(): true");
 
-            KeyStore clientKeyStore = KeyStore.getInstance("PKCS12");
+            KeyStore clientKeyStore = KeyStore.getInstance("PKCS12", Security.getProvider("BC"));
 
             clientKeyStore.load(options.getCaClientP12InputStream(), options.hasClientP12Password() ? options.getCaClientP12Password().toCharArray() : new char[0]);
             kmf.init(clientKeyStore, options.hasClientP12Password() ? options.getCaClientP12Password().toCharArray() : new char[0]);

--- a/project/gradle/libs.versions.toml
+++ b/project/gradle/libs.versions.toml
@@ -45,6 +45,7 @@ materialdrawer = "6.1.2"
 preferencex = "1.1.0"
 r8 = "3.3.75"
 guava = "31.1-jre"
+bouncycastle = "1.72"
 
 [plugins]
 android-application = { id = "com.android.application", version.ref = "agp" }
@@ -95,7 +96,6 @@ androidx-test-orchestrator = { module = "androidx.test:orchestrator", version.re
 androidx-test-espresso-idling = { module = "androidx.test.espresso:espresso-idling-resource", version.ref = "androidx-test-espresso" }
 
 osmdroid = { module = "org.osmdroid:osmdroid-android", version.ref = "osmdroid" }
-conscrypt = { module = "org.conscrypt:conscrypt-android", version.ref = "conscrypt" }
 google-material = { module = "com.google.android.material:material", version.ref = "google-material" }
 eventbus = { module = "org.greenrobot:eventbus", version.ref = "eventbus" }
 eventbus-annotationprocessor = { module = "org.greenrobot:eventbus-annotation-processor", version.ref = "eventbus" }
@@ -118,6 +118,9 @@ square-leakcanary = { module = "com.squareup.leakcanary:leakcanary-android-instr
 
 r8 = { module = "com.android.tools:r8", version.ref = "r8" }
 guava = { module = "com.google.guava:guava", version.ref = "guava" }
+
+conscrypt = { module = "org.conscrypt:conscrypt-android", version.ref = "conscrypt" }
+bouncycastle = { module = "org.bouncycastle:bcpkix-jdk15to18", version.ref = "bouncycastle" }
 
 [bundles]
 jackson = ["jackson-core", "jackson-databind", "jackson-module-kotlin", "jackson-datatype-jsr310", "jackson-threetenbp", "threetenbp"]

